[RFC] Use `meta.tests` to link from packages to the tests that test them

Ekleog opened the pull request:
Rationale
---------

Currently, tests are hard to discover. For instance, someone updating
`dovecot` might not notice that the interaction of `dovecot` with
`opensmtpd` is handled in the `opensmtpd.nix` test.

And even for someone updating `opensmtpd`, it requires manual work to go
check in `nixos/tests` whether there is actually a test, especially
given not so many packages in `nixpkgs` have tests and this is thus most
of the time useless.

Finally, for the reviewer, it is much easier to check that the “Tested
via one or more NixOS test(s)” has been checked if the file modified
already includes the list of relevant tests.

Implementation
--------------

Currently, this commit only adds the metadata in the package. Each
element of the `meta.tests` attribute is a derivation, that, when it
builds successfully, means the test has passed (ie. following the same
convention as NixOS tests).

Future Work
-----------

In the future, the tools could be made aware of this `meta.tests`
attribute, and for instance a `--with-tests` could be added to
`nix-build` so that it also builds all the tests. Or a `--without-tests`
to build without all the tests. @Profpatsch described in his NixCon talk
such systems.

Another thing that would help in the future would be the possibility to
reasonably easily have cross-derivation nix tests without the whole
NixOS VM stack. @7c6f434c already proposed such a system.

This RFC currently handles none of these concerns. Only the addition of
`meta.tests` as metadata to be used by maintainers to remember to run
relevant tests.

Still to do before merge
-----------------

- [x] Use attribute paths instead of raw imports with `../../..` (likely requires a refactoring of `release.nix` to put it inside the fix-point)
- [x] Add documentation
- [x] Add metadata in the `meta.tests` checker like @Profpatsch's [code](https://github.com/Profpatsch/nixpkgs/commit/d74203a73651d2d404a4da5ae9b0bf87d5d4218d#diff-096a3f0fd04e519accb1cde6c483e72bR179)

Ekleog committed:
dovecot, opensmtpd: add link to test in `meta.tests`

Rationale
---------

Currently, tests are hard to discover. For instance, someone updating
`dovecot` might not notice that the interaction of `dovecot` with
`opensmtpd` is handled in the `opensmtpd.nix` test.

And even for someone updating `opensmtpd`, it requires manual work to go
check in `nixos/tests` whether there is actually a test, especially
given not so many packages in `nixpkgs` have tests and this is thus most
of the time useless.

Finally, for the reviewer, it is much easier to check that the “Tested
via one or more NixOS test(s)” has been checked if the file modified
already includes the list of relevant tests.

Implementation
--------------

Currently, this commit only adds the metadata in the package. Each
element of the `meta.tests` attribute is a derivation that, when it
builds successfully, means the test has passed (ie. following the same
convention as NixOS tests).

Future Work
-----------

In the future, the tools could be made aware of this `meta.tests`
attribute, and for instance a `--with-tests` could be added to
`nix-build` so that it also builds all the tests. Or a `--without-tests`
to build without all the tests. @Profpatsch described in his NixCon talk
such systems.

Another thing that would help in the future would be the possibility to
reasonably easily have cross-derivation nix tests without the whole
NixOS VM stack. @7c6f434c already proposed such a system.

This RFC currently handles none of these concerns. Only the addition of
`meta.tests` as metadata to be used by maintainers to remember to run
relevant tests.

samueldr commented:
Any trick in mind to make it more palatable than a path full of `../` that will invariably change depending on the depth of the derivation? [The tests are already being put in an attrset for release](https://github.com/NixOS/nixpkgs/blob/cb4263a7a8a70af53aad181ad13c37485cab65e6/nixos/release.nix#L243-L421), maybe it could be moved into a more reusable place?

Ekleog commented:
Hmm… From what I understand, it'd require a refactoring from `release.nix`, as that's not readily available from the fixed point? If so, maybe doing it once the overall design is more or less consensual would be better, in order for the refactoring not to hide the important parts during discussion :) Thank you for the idea!

dotlambda commented on pkgs/servers/mail/dovecot/default.nix:
```diff
@@ -74,5 +74,8 @@ stdenv.mkDerivation rec {
     maintainers = with stdenv.lib.maintainers; [ peti rickynils fpletz ];
     platforms = stdenv.lib.platforms.unix;
+    tests = [
+      (import ../../../../nixos/tests/opensmtpd.nix {})
```
There should also be a link to `nixos/tests/dovecot.nix`.

Ekleog replied:
Indeed, I had only picked the `opensmtpd` test I had written as an example for the purpose of the RFC, all tests should be added to all relevant packages, I think so too :) There appear to be ~200 of them, though, so I'm not sure all of them could reasonably go into a single PR without risking to have bitrotten before even being completed?

aszlig commented:
I'm not quite sure whether this really should belong together, because the NixOS VM tests are more tailured towards the NixOS service, so tests should ideally be linked to the NixOS module (possibly in a way so you can easily say "run all tests for the current configuration", prototype [here](https://github.com/openlab-aux/vuizvui/blob/3b3f645bbf6f55cb19a1aac3374940f4fa91ddfb/modules/core/tests.nix)) rather than to the package. If the test is really a test for the package itself (like unit tests etc.), then something like this sounds more appropriate.

If this is only for running tests on pull requests, this sounds better, but I also don't like the `../../...`, maybe use a list of attribute paths instead?

aszlig commented:
Using attribute paths btw. also has the benefit that `nix-instantiate --eval --strict '<nixpkgs>' -A dovecot.meta` will work, IIRC something like this was (or is) used in Hydra.

7c6f434c commented:
@aszlig if some package is disproportionately likely to break a NixOS test, we might want to schedule that test anyway.

If the test is executed in a background batch together with some rebuilds, and the test is already debugged for other reasons, VM overhead is less of an issue.

Profpatsch commented:
We already had a type cobbled together that these tests should have:
https://github.com/Profpatsch/nixpkgs/commit/d74203a73651d2d404a4da5ae9b0bf87d5d4218d#diff-096a3f0fd04e519accb1cde6c483e72bR179

In plain words, `tests` should be an attribute set that contains tests, where a test is a derivation that has a passthrough boolean field `isVmTest`. That’s just a first idea of course.

Ekleog commented:
@7c6f434c I think the second commit of this PR (added after I noticed the ofborg check failed) should do the “updating the validation scripts” part :)

---

@aszlig To me, the point of `meta.tests` is to be run before accepting a package update into `nixpkgs`, indeed, given unit tests are (I think?) supposed to be run in the `checkPhase`. The aim is mostly to allow a single test (like here the OpenSMTPD test) to serve for two derivations (here, `opensmtpd` and `dovecot`) :)

I think a scheme, like you put forward, of tying tests to a NixOS configuration, would be useful too, but I'm not sure it'd fit the same aim, as your scheme appears to be more oriented towards individuals accepting (or not) a channel bump as an upgrade :)

Basically, here I'm using a NixOS test because that's all I have to smoke-test that OpenSMTPD / dovecot appear to be working relatively consistently (ie. are able to talk to each over on something that at least look like resp. SMTP and binary-MDA). It's not particularly tied to NixOS in my opinion, but NixOS is (currently) the only testing framework that we have :)

---

@Profpatsch Your proposal appears much more complete than mine! would you rather I close this PR and you open one about it, or I include your changes into mine?

Also, there are still one point I'm not fully comfortable with with your `check-meta` update (apart from the `foo` attribute, that I guess is a WIP test element?): `isVmTest`.

There are two types of test I can think of off the top of my head that could be used right now:
1. NixOS tests
2. Dependencies (eg. a library setting a light dependency as a smoke test to check it hasn't completely broken its API)

And potentially in the future, there could be a testing scheme that doesn't require a full-blown NixOS system, like the one @7c6f434c proposed (test category 3).

I think the filtering could be done along two axes:
1. How long does it take to run the build? (the equivalent of your `isVmTest`, I guess, but test category 3 could also potentially take long)
2. Should tests be recursive? (ie. should the tests of the test packages be executed? this could be useful for test category 2, eg. `openssl` could declare `curl` as a recursive-test, and `curl` would have a test that checks it against an https `nginx` -- this would smoke-test that the API of `openssl` appears not to have broken enough to break `curl`)

I'm not completely sure about case 2 (it could be done by directly including the test case for `curl` against `nginx` for instance), but I think anyway `isVmTest` could (should?) be renamed `isSlowTest`.

What do you think about this? :)

---

I've added the still-to-do points to the top post, will try to get some done tomorrow :)

Profpatsch commented:
> What do you think about this? :)

I think we should basically have an open record of properties for tests.

```
let
  testProperties = productOpt {
    open = true;
    req = {
      # times mostly defined by test runners, comments are proposals
      # test runners might decide to cap after certain time
      runningTime = sum {
        short = unit; # under 1 minute
        medium = unit; # under 5 minutes
        long = unit; # under 15 minutes
        unbound = unit; # more than 15 minutes
        maxMinutes = int;
      };

      needsVMSupport = bool;
    };
  };
in testProperties
```

These options are then used by the test runner to decide which tests to run and how long the run can last until it is aborted. More options might be added later.

GrahamcOfBorg commented:
Success on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.44439&attempt_id=cb098e51-021c-4f06-87df-372d8617460d)

Attempted: dovecot, opensmtpd

<details><summary>Partial log (click to expand)</summary><p>

```
copying path '/nix/store/qksj8mx1cdvhhblqxshjzfri4yc3qrdb-libevent-2.1.8' from 'https://cache.nixos.org'...
copying path '/nix/store/ljs86xdssnqnqvpa6xnyhvkzfv6w4prg-libsodium-1.0.16' from 'https://cache.nixos.org'...
copying path '/nix/store/29j6kf19lxj81hyb6mbzqp42w9s47vx9-lz4-1.8.2' from 'https://cache.nixos.org'...
copying path '/nix/store/gd222hhrhwm2rjzd7c4vi55vfxrrg5zk-cyrus-sasl-2.1.26' from 'https://cache.nixos.org'...
copying path '/nix/store/v2hsgafk6wgrbgzrai05wnncwf9dwkzm-libtool-2.4.6-lib' from 'https://cache.nixos.org'...
copying path '/nix/store/wj214jamqiysbkw5dbvpgyvdnmlpkahr-openldap-2.4.45' from 'https://cache.nixos.org'...
copying path '/nix/store/vi9n4cszz9rqfmf3p5c8chvqv2pviywb-opensmtpd-6.0.3p1' from 'https://cache.nixos.org'...
copying path '/nix/store/8dyq4hg5p6c074kcxbrcl34miwg6n5bf-dovecot-2.3.2.1' from 'https://cache.nixos.org'...
/nix/store/8dyq4hg5p6c074kcxbrcl34miwg6n5bf-dovecot-2.3.2.1
/nix/store/vi9n4cszz9rqfmf3p5c8chvqv2pviywb-opensmtpd-6.0.3p1
```
</p></details>

GrahamcOfBorg commented:
Success on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.44439&attempt_id=0e5044cb-07c9-4efb-9b33-fe670480768a)

Attempted: dovecot, opensmtpd

<details><summary>Partial log (click to expand)</summary><p>

```
  /nix/store/0v05paacys9bnb9fxvvhwk8j5mqg34lw-dovecot-2.3.2.1
  /nix/store/5h9kpirrh5fw9imblwxm4d1lr8acwsk6-clucene-core-2.3.3.4
  /nix/store/llyvxfx789yrwbljy3pig9q7al64bg5k-libasr-1.0.2
  /nix/store/lmjpbp2vh6k6via7lyb92zxjfb3gpw7j-opensmtpd-6.0.3p1
copying path '/nix/store/5h9kpirrh5fw9imblwxm4d1lr8acwsk6-clucene-core-2.3.3.4' from 'https://cache.nixos.org'...
copying path '/nix/store/llyvxfx789yrwbljy3pig9q7al64bg5k-libasr-1.0.2' from 'https://cache.nixos.org'...
copying path '/nix/store/lmjpbp2vh6k6via7lyb92zxjfb3gpw7j-opensmtpd-6.0.3p1' from 'https://cache.nixos.org'...
copying path '/nix/store/0v05paacys9bnb9fxvvhwk8j5mqg34lw-dovecot-2.3.2.1' from 'https://cache.nixos.org'...
/nix/store/0v05paacys9bnb9fxvvhwk8j5mqg34lw-dovecot-2.3.2.1
/nix/store/lmjpbp2vh6k6via7lyb92zxjfb3gpw7j-opensmtpd-6.0.3p1
```
</p></details>

GrahamcOfBorg commented:
Success on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.44439&attempt_id=e8a5e784-d3f8-4b6e-9c3b-f9b6e98fb11d)

Attempted: dovecot, opensmtpd

<details><summary>Partial log (click to expand)</summary><p>

```
/nix/store/8dyq4hg5p6c074kcxbrcl34miwg6n5bf-dovecot-2.3.2.1
/nix/store/vi9n4cszz9rqfmf3p5c8chvqv2pviywb-opensmtpd-6.0.3p1
```
</p></details>

GrahamcOfBorg commented:
Success on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.44439&attempt_id=f0fa5969-8323-4093-9757-2fbda59b0b07)

Attempted: dovecot, opensmtpd

<details><summary>Partial log (click to expand)</summary><p>

```
  /nix/store/0v05paacys9bnb9fxvvhwk8j5mqg34lw-dovecot-2.3.2.1
  /nix/store/5h9kpirrh5fw9imblwxm4d1lr8acwsk6-clucene-core-2.3.3.4
  /nix/store/llyvxfx789yrwbljy3pig9q7al64bg5k-libasr-1.0.2
  /nix/store/lmjpbp2vh6k6via7lyb92zxjfb3gpw7j-opensmtpd-6.0.3p1
copying path '/nix/store/5h9kpirrh5fw9imblwxm4d1lr8acwsk6-clucene-core-2.3.3.4' from 'https://cache.nixos.org'...
copying path '/nix/store/llyvxfx789yrwbljy3pig9q7al64bg5k-libasr-1.0.2' from 'https://cache.nixos.org'...
copying path '/nix/store/0v05paacys9bnb9fxvvhwk8j5mqg34lw-dovecot-2.3.2.1' from 'https://cache.nixos.org'...
copying path '/nix/store/lmjpbp2vh6k6via7lyb92zxjfb3gpw7j-opensmtpd-6.0.3p1' from 'https://cache.nixos.org'...
/nix/store/0v05paacys9bnb9fxvvhwk8j5mqg34lw-dovecot-2.3.2.1
/nix/store/lmjpbp2vh6k6via7lyb92zxjfb3gpw7j-opensmtpd-6.0.3p1
```
</p></details>

GrahamcOfBorg commented:
Success on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.44439&attempt_id=9f68b77e-0114-40fb-923e-d4c39c113778)

Attempted: dovecot, opensmtpd

<details><summary>Partial log (click to expand)</summary><p>

```
shrinking /nix/store/zjmkciznnc2n4vvkpnq7zqvwv25iry41-dovecot-2.3.2.1/sbin/dovecot
gzipping man pages under /nix/store/zjmkciznnc2n4vvkpnq7zqvwv25iry41-dovecot-2.3.2.1/share/man/
strip is /nix/store/1hi76hr87bd1y1q1qjk0lv8nmcjip1c8-binutils-2.30/bin/strip
stripping (with command strip and flags -S) in /nix/store/zjmkciznnc2n4vvkpnq7zqvwv25iry41-dovecot-2.3.2.1/lib  /nix/store/zjmkciznnc2n4vvkpnq7zqvwv25iry41-dovecot-2.3.2.1/libexec  /nix/store/zjmkciznnc2n4vvkpnq7zqvwv25iry41-dovecot-2.3.2.1/bin  /nix/store/zjmkciznnc2n4vvkpnq7zqvwv25iry41-dovecot-2.3.2.1/sbin
patching script interpreter paths in /nix/store/zjmkciznnc2n4vvkpnq7zqvwv25iry41-dovecot-2.3.2.1
/nix/store/zjmkciznnc2n4vvkpnq7zqvwv25iry41-dovecot-2.3.2.1/libexec/dovecot/decode2text.sh: interpreter directive changed from "/bin/sh" to "/nix/store/dkh7l9a4sx7zqh8riqbj3z21sz25p8xy-bash-4.4-p23/bin/sh"
checking for references to /build in /nix/store/zjmkciznnc2n4vvkpnq7zqvwv25iry41-dovecot-2.3.2.1...
moving /nix/store/zjmkciznnc2n4vvkpnq7zqvwv25iry41-dovecot-2.3.2.1/sbin/* to /nix/store/zjmkciznnc2n4vvkpnq7zqvwv25iry41-dovecot-2.3.2.1/bin
/nix/store/zjmkciznnc2n4vvkpnq7zqvwv25iry41-dovecot-2.3.2.1
/nix/store/lmjpbp2vh6k6via7lyb92zxjfb3gpw7j-opensmtpd-6.0.3p1
```
</p></details>

GrahamcOfBorg commented:
Success on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.44439&attempt_id=eb1173b6-c323-4652-8c0f-624fb926bc26)

Attempted: dovecot, opensmtpd

<details><summary>Partial log (click to expand)</summary><p>

```
shrinking /nix/store/z17hyaagddgxlpncxmqk6a70hn2kh27d-dovecot-2.3.2.1/bin/doveadm
gzipping man pages under /nix/store/z17hyaagddgxlpncxmqk6a70hn2kh27d-dovecot-2.3.2.1/share/man/
strip is /nix/store/zrs21zqcchgyabjf4xfimncdq16njizc-binutils-2.30/bin/strip
stripping (with command strip and flags -S) in /nix/store/z17hyaagddgxlpncxmqk6a70hn2kh27d-dovecot-2.3.2.1/lib  /nix/store/z17hyaagddgxlpncxmqk6a70hn2kh27d-dovecot-2.3.2.1/libexec  /nix/store/z17hyaagddgxlpncxmqk6a70hn2kh27d-dovecot-2.3.2.1/bin  /nix/store/z17hyaagddgxlpncxmqk6a70hn2kh27d-dovecot-2.3.2.1/sbin
patching script interpreter paths in /nix/store/z17hyaagddgxlpncxmqk6a70hn2kh27d-dovecot-2.3.2.1
/nix/store/z17hyaagddgxlpncxmqk6a70hn2kh27d-dovecot-2.3.2.1/libexec/dovecot/decode2text.sh: interpreter directive changed from "/bin/sh" to "/nix/store/6vqgi9d49smsbr2qxra52yhipg0yxf9f-bash-4.4-p23/bin/sh"
checking for references to /build in /nix/store/z17hyaagddgxlpncxmqk6a70hn2kh27d-dovecot-2.3.2.1...
moving /nix/store/z17hyaagddgxlpncxmqk6a70hn2kh27d-dovecot-2.3.2.1/sbin/* to /nix/store/z17hyaagddgxlpncxmqk6a70hn2kh27d-dovecot-2.3.2.1/bin
/nix/store/z17hyaagddgxlpncxmqk6a70hn2kh27d-dovecot-2.3.2.1
/nix/store/vi9n4cszz9rqfmf3p5c8chvqv2pviywb-opensmtpd-6.0.3p1
```
</p></details>

Profpatsch commented on pkgs/top-level/all-packages.nix:
```diff
@@ -71,6 +71,25 @@ with pkgs;
+
+  nixosTests =
+    let
+      system = builtins.currentSystem;
```
This doesn’t work, I think. cc @Ericson2314

Ekleog replied:
I don't really know: NixOS test derivations are supposed to be built on the local system, which I'd guess to be `builtins.currentSystem`? Not sure at all though, so cc @Ericson2314 indeed :) I.must find it hard to imagine what's the cross-derivation story of NIxOS tests.

Ekleog replied:
@Ericson2314 ping :) (2 months sounds like a reasonable waiting time :p)

Profpatsch reviewed (changes requested):
Not sure, but I think this needs to use some cross-stuff.

Ekleog commented:
Absent any reply from @Ericson2314 about whether `builtins.currentSystem` is correctly used here and given it can certainly be fixed after the fact if it is not, I added a TODO item, squashed and rebased on `master`.

I have tested that `nix-build . -A opensmtpd.meta.tests -A dovecot.meta.tests` correctly passes, as well as that the manual builds and displays correctly.

Given the absence of unresolved issues, I think this is ready to move forward :)

GrahamcOfBorg commented:
<!--REQUEST_ID=41e9c6ad-867b-41c8-96a3-782319014fda-->
Success on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.44439&attempt_id=9c104b26-b6ef-4d86-87be-3b1c6beb7d3e)

Attempted: dovecot, opensmtpd

<details><summary>Partial log (click to expand)</summary><p>

```
shrinking /nix/store/7gj4a8m6zm6rg4x149vq73bzijpd8nza-opensmtpd-6.4.0p1/libexec/opensmtpd/mail.maildir
shrinking /nix/store/7gj4a8m6zm6rg4x149vq73bzijpd8nza-opensmtpd-6.4.0p1/libexec/opensmtpd/mail.lmtp
shrinking /nix/store/7gj4a8m6zm6rg4x149vq73bzijpd8nza-opensmtpd-6.4.0p1/sbin/smtpctl
shrinking /nix/store/7gj4a8m6zm6rg4x149vq73bzijpd8nza-opensmtpd-6.4.0p1/sbin/smtpd
gzipping man pages under /nix/store/7gj4a8m6zm6rg4x149vq73bzijpd8nza-opensmtpd-6.4.0p1/share/man/
strip is /nix/store/p9akxn2sfy4wkhqdqa3li97pc6jaz3r1-binutils-2.30/bin/strip
stripping (with command strip and flags -S) in /nix/store/7gj4a8m6zm6rg4x149vq73bzijpd8nza-opensmtpd-6.4.0p1/libexec  /nix/store/7gj4a8m6zm6rg4x149vq73bzijpd8nza-opensmtpd-6.4.0p1/bin  /nix/store/7gj4a8m6zm6rg4x149vq73bzijpd8nza-opensmtpd-6.4.0p1/sbin
patching script interpreter paths in /nix/store/7gj4a8m6zm6rg4x149vq73bzijpd8nza-opensmtpd-6.4.0p1
checking for references to /build in /nix/store/7gj4a8m6zm6rg4x149vq73bzijpd8nza-opensmtpd-6.4.0p1...
moving /nix/store/7gj4a8m6zm6rg4x149vq73bzijpd8nza-opensmtpd-6.4.0p1/sbin/* to /nix/store/7gj4a8m6zm6rg4x149vq73bzijpd8nza-opensmtpd-6.4.0p1/bin
```
</p></details>

GrahamcOfBorg commented:
<!--REQUEST_ID=41e9c6ad-867b-41c8-96a3-782319014fda-->
Success on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.44439&attempt_id=7bb8db4b-7759-4c98-a133-c0b8ca7c2357)

Attempted: dovecot, opensmtpd

<details><summary>Partial log (click to expand)</summary><p>

```
shrinking /nix/store/3bv0id3xags47ayczfy1rwc6aq4bfji0-opensmtpd-6.4.0p1/libexec/opensmtpd/mail.mda
shrinking /nix/store/3bv0id3xags47ayczfy1rwc6aq4bfji0-opensmtpd-6.4.0p1/libexec/opensmtpd/mail.maildir
shrinking /nix/store/3bv0id3xags47ayczfy1rwc6aq4bfji0-opensmtpd-6.4.0p1/libexec/opensmtpd/mail.local
shrinking /nix/store/3bv0id3xags47ayczfy1rwc6aq4bfji0-opensmtpd-6.4.0p1/libexec/opensmtpd/mail.lmtp
gzipping man pages under /nix/store/3bv0id3xags47ayczfy1rwc6aq4bfji0-opensmtpd-6.4.0p1/share/man/
strip is /nix/store/vcc4svb8gy29g4pam2zja6llkbcwsyiq-binutils-2.30/bin/strip
stripping (with command strip and flags -S) in /nix/store/3bv0id3xags47ayczfy1rwc6aq4bfji0-opensmtpd-6.4.0p1/libexec  /nix/store/3bv0id3xags47ayczfy1rwc6aq4bfji0-opensmtpd-6.4.0p1/bin  /nix/store/3bv0id3xags47ayczfy1rwc6aq4bfji0-opensmtpd-6.4.0p1/sbin
patching script interpreter paths in /nix/store/3bv0id3xags47ayczfy1rwc6aq4bfji0-opensmtpd-6.4.0p1
checking for references to /build in /nix/store/3bv0id3xags47ayczfy1rwc6aq4bfji0-opensmtpd-6.4.0p1...
moving /nix/store/3bv0id3xags47ayczfy1rwc6aq4bfji0-opensmtpd-6.4.0p1/sbin/* to /nix/store/3bv0id3xags47ayczfy1rwc6aq4bfji0-opensmtpd-6.4.0p1/bin
```
</p></details>

timokau commented on doc/meta.xml:
```diff
@@ -250,6 +250,61 @@ meta.platforms = stdenv.lib.platforms.linux;
+{
+  # ...
+  meta.tests = {
+    basic-functionality-and-dovecot-integration = nixosTests.opensmtpd;
```
Not sure if this has been discussed before (can't find it with a quick skim), but is there a reason to make this a set instead of a simple list? I feel like having to give every test a new name only reduces redundancy. If there is additional information (like what exactly is tested in this case), maybe renaming the original test or a comment would be better?

Ekleog replied:
I had originally implemented this as a list, and @Profpatsch's comment (https://github.com/NixOS/nixpkgs/pull/44439#issuecomment-410447611) made me notice it could be written as an attrset. At the time, it was a good thing, because tests needed to be `import`-ed, and the additional parenthesis in lists are often a source of errors.

Now that the tests are easily accessible from the fixpoint… I don't have any strong sentiment about it. @Profpatsch maybe has?

Profpatsch replied:
A list is kinda hard to work with, it can only be mapped over and replaced by an other list. There is no (sane) way to reference single elements in it. So if there should ever be the need to work with one of these tests especially (e.g. for overriding or removing it), you don’t have a good reference to work with (and no, the test’s name attribute is not a good reference).

timokau replied:
I still don't like the "UX" of attribute sets here very much, but your argument makes sense. Using `inherit` would at least remove the redundancy.

So as a summary I'm not quite happy but have no better idea.

Profpatsch commented:
Not from my side.

7c6f434c commented:
I wonder if we ever find a way to specify timeouts properly (to express a linear combination of some calibration build, maybe a minimal VM boot time, wall-clock seconds and maybe something more) without losing our minds. But of course it is not a blocker here.

Profpatsch commented:
> I wonder if we ever find a way to specify timeouts properly

It should be worth a look to see [how bazel does it](https://docs.bazel.build/versions/master/be/common-definitions.html#common-attributes-tests), and also other build tools that support testsi.

Ericson2314 commented:
Sorry I diddn't see this. `builtins.currentSystem` is very incorrect, both for cross and for hydra which evaluates different native builds.

Ericson2314 commented:
It's not letting me comment inline for some reason, but `nixpkgs = pkgs` is also wrong. `nixpkgs` is supposed to be unimported nixpkgs, which `release.nix` will them import various ways. Really all relating to the NixOS tests needs some refactoring, both for correctness, and for performance in not evaluating nixpkgs too many times. The best thing to do is separate the tests from `release.nix`

Ekleog commented:
@Ericson2314 I've opened https://github.com/NixOS/nixpkgs/pull/50028 to (hopefully) handle your first comment.

The second appears like a much larger refactoring. Do you think we should go back on this change until we get it exactly right, or is it “good enough” for now and can it be improved upon once we manage to do the refactoring?

FRidh commented:
Why is the evaluator complaining:
```
derivation 'dovecot-2.3.3' has invalid meta attribute 'tests'
derivation 'opensmtpd-6.4.0p2' has invalid meta attribute 'tests'
```

Ekleog commented:
@matthewbauer Yes, that's the idea, along with having ofborg automatically build these :) Do you think [here](https://github.com/Ekleog/nixpkgs/blob/02e1f00ffdd0c1a74fbc0be5f314b8921b0c9165/doc/meta.xml#L253-L280) would be a good place to add this `nix-build` command?

@FRidh Where are you seeing that? I've noticed ofborg complain about [missing attributes](https://gist.github.com/GrahamcOfBorg/7e191b96cdb9c1bfe66e8abcdf594bfa) (for tests that don't exist in the system, though I don't know whether we actually have those in practice and I hit this only when trying to fix the cross-compilation story), but can't find where this happens for `'tests'`, and it should be taken care of by [this part of the PR](https://github.com/NixOS/nixpkgs/pull/44439/files#diff-096a3f0fd04e519accb1cde6c483e72bR169) :/

FRidh commented:
@Ekleog run `exec nix-env -f . -qaP --meta --json --drv-path --show-trace >/dev/null` on `master` or `staging-next`.

peti commented:
> unknown meta entries get reported as typos by various tools

Unknown `meta` attributes are reported as a warning by *Nix* itself. See https://github.com/NixOS/nixpkgs/issues/50230.

grahamc commented:
@peti perhaps we should revert this entire PR, then?

grahamc commented on doc/meta.xml:
```diff
@@ -250,6 +250,61 @@ meta.platforms = stdenv.lib.platforms.linux;
+      An attribute set with as values tests. A test is a derivation, which
+      builds successfully when the test passes, and fails to build otherwise. A
+      derivation that is a test requires some <literal>meta</literal> elements
+      to be defined: <literal>needsVMSupport</literal> (automatically filled-in
```
This should probably not be created, since this is already covered by the requiredFeatures of the derivation.

Ekleog commented:
@Ericson2314 @FRidh @peti I've opened https://github.com/NixOS/nixpkgs/pull/50233, that should hopefully handle all the remaining concerns.

Ekleog commented:
It has already been reverted by @peti, so there is as much time as need be :) (well, to be more precise the `meta.tests` lines have been commented out)

Unless you mean reverting the whole PR and doing it again in the fixing PR? I don't think I can manage a clean reviewable history with that, compared to what I have currently made: it'd mean the three atomic “drop meta.needsVMSupport” (+3-19), “rename into passthru.tests” (+20-14, mostly comments) and “use the newly-extracted all-tests.nix” (+5-17) commits would get squashed into a jumbo “add meta.tests” that'd include this PR (+98-4), but I can do if you think it'd be easier for you to review this way.
